src/engraving/dom/harmony.cpp

```diff
@@ -889,566 +1055,633 @@ Fraction Harmony::ticksTillNext(int utick, bool stopAtMeasureEnd) const
+    }
+
+    // Check all symbols in string are available in this font
+    FontMetrics fm(f);
+    bool fail = false;
+    for (size_t i = 0; i < m_text.size(); ++i) {
+        const Char& c = m_text.at(i);
+        if (c.isHighSurrogate()) {
+            if (i + 1 == m_text.size()) {
+                ASSERT_X("bad string");
+            }
+            const Char& c2 = m_text.at(i + 1);
+            ++i;
+            char32_t v = Char::surrogateToUcs4(c, c2);
+            if (!fm.inFontUcs4(v)) {
+                fail = true;
+                break;
+            }
+        } else {
+            if (!fm.inFont(c)) {
+                fail = true;
+                break;
+            }
+        }
+    }
+    // Fallback to default musical text font
+    if (fail) {
+        m_font.setFamily(String::fromUtf8(FALLBACK_SYMBOLTEXT_FONT), Font::Type::MusicSymbolText);
     }
-    return cd;
 }
 
 //---------------------------------------------------------
-//   getRealizedHarmony
-//    get realized harmony or create one for the current symbol
-//    also updates the realized harmony and accounts for
-//    transposition. RealizedHarmony objects cannot be cached
-//    since the notes generated depends on context rather than
-//    just root, bass, chord symbol, and voicing.
+//   render
 //---------------------------------------------------------
 
-const RealizedHarmony& Harmony::getRealizedHarmony() const
+void Harmony::render(const String& s, HarmonyRenderCtx& ctx)
 {
-    const Staff* st = staff();
-    IF_ASSERT_FAILED(st) {
-        return m_realizedHarmony;
+    if (s.isEmpty()) {
+        return;
     }
 
-    const Fraction tick = this->tick();
-    const CapoParams& capo = st->capo(tick);
+    Font f = m_harmonyType != HarmonyType::ROMAN ? m_fontList.front() : font();
+    TextSegment* ts = new TextSegment(s, f, ctx.x(), ctx.y(), ctx.hAlign);
+    ctx.textList.push_back(ts);
+    ctx.movex(ts->width());
+}
 
-    int offset = 0;
-    if (capo.active) {
-        offset = capo.fretPosition;
+void Harmony::render(SymId sym, HarmonyRenderCtx& ctx)
+{
+    if (sym == SymId::noSym) {
+        return;
     }
 
-    Interval interval = st->part()->instrument(tick)->transpose();
-    if (!style().styleB(Sid::concertPitch)) {
-        offset += interval.chromatic;
+    Font f = m_harmonyType != HarmonyType::ROMAN ? m_fontList.front() : font();
+    f.setFamily(style().styleSt(Sid::musicalTextFont), Font::Type::MusicSymbolText);
+
+    String s = score()->engravingFont()->toString(sym);
+
+    TextSegment* ts = new TextSegment(s, f, ctx.x(), ctx.y(), ctx.hAlign);
+    ctx.textList.push_back(ts);
+    ctx.movex(ts->width());
+}
+
+//---------------------------------------------------------
+//   render
+//---------------------------------------------------------
+
+void Harmony::render(const std::list<RenderAction*>& renderList, HarmonyRenderCtx& ctx, int tpc, NoteSpellingType noteSpelling,
+                     NoteCaseType noteCase, double noteMag)
```

Comment:
Again, not for this PR, but these functions would be better moved to a Layout submodule